helix-vcs/src/diff.rs

```diff
@@ -0,0 +1,186 @@
+use std::sync::Arc;
+
+use helix_core::Rope;
+use imara_diff::Algorithm;
+use parking_lot::{Mutex, MutexGuard};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedSender};
+use tokio::sync::{Notify, RwLock};
+use tokio::task::JoinHandle;
+
+use crate::diff::worker::DiffWorker;
+
+mod line_cache;
+mod worker;
+
+type RedrawHandle = Arc<(Notify, RwLock<()>)>;
+
+// The order of enum variants is used by the PartialOrd
+// derive macro, DO NOT REORDER
+#[derive(PartialEq, PartialOrd)]
+enum RenderStrategy {
+    Async,
+    SyncWithTimeout,
+    Sync,
+}
+
+struct Event {
+    text: Rope,
+    is_base: bool,
+    render_strategy: RenderStrategy,
+}
+
+#[derive(Clone, Debug)]
+pub struct DiffHandle {
+    channel: UnboundedSender<Event>,
+    hunks: Arc<Mutex<Vec<Hunk>>>,
+    inverted: bool,
+}
+
+impl DiffHandle {
+    pub fn new(diff_base: Rope, doc: Rope, redraw_handle: RedrawHandle) -> DiffHandle {
+        DiffHandle::new_with_handle(diff_base, doc, redraw_handle).0
+    }
+
+    fn new_with_handle(
+        diff_base: Rope,
+        doc: Rope,
+        redraw_handle: RedrawHandle,
+    ) -> (DiffHandle, JoinHandle<()>) {
+        let (sender, receiver) = unbounded_channel();
+        let hunks: Arc<Mutex<Vec<Hunk>>> = Arc::default();
+        let worker = DiffWorker {
+            channel: receiver,
+            hunks: hunks.clone(),
+            new_hunks: Vec::default(),
+            redraw_handle,
+            difff_finished_notify: Arc::default(),
+        };
+        let handle = tokio::spawn(worker.run(diff_base, doc));
+        let differ = DiffHandle {
+            channel: sender,
+            hunks,
+            inverted: false,
+        };
+        (differ, handle)
+    }
+
+    pub fn invert(&mut self) {
+        self.inverted = !self.inverted;
+    }
+
+    pub fn hunks(&self) -> FileHunks {
+        FileHunks {
+            hunks: self.hunks.lock(),
+            inverted: self.inverted,
+        }
+    }
+
+    pub fn update_document(&self, doc: Rope, block: bool) -> bool {
+        let mode = if block {
+            RenderStrategy::Sync
+        } else {
+            RenderStrategy::SyncWithTimeout
+        };
+        self.update_document_impl(doc, self.inverted, mode)
+    }
+
+    pub fn update_diff_base(&self, diff_base: Rope) -> bool {
+        self.update_document_impl(diff_base, !self.inverted, RenderStrategy::Async)
+    }
+
+    fn update_document_impl(&self, text: Rope, is_base: bool, mode: RenderStrategy) -> bool {
+        let event = Event {
+            text,
+            is_base,
+            render_strategy: mode,
+        };
+        self.channel.send(event).is_ok()
+    }
+}
+
+// TODO configuration
+/// synchronous debounce value should be low
+/// so we can update synchronously most of the time
+const DIFF_DEBOUNCE_TIME_SYNC: u64 = 1;
+/// maximum time that rendering should be blocked until the diff finishes
+const SYNC_DIFF_TIMEOUT: u64 = 50;
+const DIFF_DEBOUNCE_TIME_ASYNC: u64 = 100;
```

Comment:
I don't think this is as critical here because it this timeout only affects fully async redraws and is not really tied to frames but I changed it to `96` anwyay